File: backend/src/modules/bale_monitor/parser.js
const INTERVAL_RE = /^(\d+)([smh])$/i;

function parseCommand(text) {
  if (typeof text !== 'string' || !text.trim().startsWith('/')) return null;
  const parts = text.trim().split(/\s+/);
  return { name: parts[0].toLowerCase(), args: parts.slice(1) };
}

function validateUrl(raw) {
  const value = String(raw || '').trim();
  if (!/^https?:\/\//i.test(value)) {
    throw new Error('URL باید با http:// یا https:// شروع شود.');
  }
  return value;
}

function parseInterval(raw, min, max) {
  const m = String(raw || '').trim().match(INTERVAL_RE);
  if (!m) throw new Error('فرمت interval نامعتبر است. مثال: 30s یا 5m یا 1h');
  const n = Number(m[1]);
  const unit = m[2].toLowerCase();
  const factor = unit === 's' ? 1 : unit === 'm' ? 60 : 3600;
  const seconds = n * factor;
  if (seconds < min || seconds > max) {
    throw new Error('Interval باید بین 10 ثانیه تا 24 ساعت باشد.');
  }
  return seconds;
}

function parseDailyTime(raw) {
  const value = String(raw || '').trim();
  if (!/^([01]\d|2[0-3]):([0-5]\d)$/.test(value)) {
    throw new Error('فرمت زمان نامعتبر است. مثال: 14:30');
  }
  return value;
}

function parseId(raw) {
  const id = Number(String(raw || '').trim());
  if (!Number.isInteger(id) || id <= 0) throw new Error('id نامعتبر است.');
  return id;
}

module.exports = { parseCommand, validateUrl, parseInterval, parseDailyTime, parseId };

File: uptime/modules/bale_monitor/parser.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
import re

from .config import MAX_INTERVAL_SECONDS, MIN_INTERVAL_SECONDS

INTERVAL_PATTERN = re.compile(r"^(\d+)([smh])$")


@dataclass(slots=True)
class ParsedCommand:
    name: str
    args: list[str]


class CommandParseError(ValueError):
    pass


def parse_command(text: str) -> ParsedCommand:
    if not text or not text.startswith("/"):
        raise CommandParseError("Command must start with '/'.")

    parts = text.strip().split()
    if not parts:
        raise CommandParseError("Empty command.")

    return ParsedCommand(name=parts[0].lower(), args=parts[1:])


def validate_url(url: str) -> str:
    clean = url.strip()
    if clean.startswith("http://") or clean.startswith("https://"):
        return clean
    raise CommandParseError("Invalid URL. It must start with http:// or https://")


def parse_interval(raw: str) -> int:
    match = INTERVAL_PATTERN.match(raw.lower().strip())
    if not match:
        raise CommandParseError("Invalid interval. Use formats like 30s, 5m, 1h")

    amount = int(match.group(1))
    unit = match.group(2)
    multiplier = {"s": 1, "m": 60, "h": 3600}[unit]
    seconds = amount * multiplier

    if seconds < MIN_INTERVAL_SECONDS or seconds > MAX_INTERVAL_SECONDS:
        raise CommandParseError("Interval must be between 10 seconds and 24 hours")

    return seconds


def parse_daily_time(raw: str) -> str:
    value = raw.strip()
    try:
        datetime.strptime(value, "%H:%M")
    except ValueError as exc:
        raise CommandParseError("Invalid time format. Use HH:MM (24-hour)") from exc
    return value


def parse_monitor_id(raw: str) -> int:
    if not raw.isdigit():
        raise CommandParseError("Monitor id must be numeric")
    return int(raw)
